src/utils/locations.ts

```diff
@@ -61,6 +75,46 @@ export function getMainFunctionTokenLocation(
   return location!;
 }
 
+// As `ReportDescriptor` may contain either `message` or `messageId` prop,
+// we force the presence of `message` property by using the following type alias.
+export type ReportDescriptor = Rule.ReportDescriptor & { message: string };
+
+/**
+ * Wrapper for `context.report`, supporting secondary locations and cost.
+ * Encode those extra information in the issue message when rule is executed
+ * in Sonar* environment.
+ */
+export function report(
+  context: Rule.RuleContext,
+  reportDescriptor: ReportDescriptor,
+  secondaryLocations: IssueLocation[] = [],
+  cost?: number,
+) {
+  const { message } = reportDescriptor;
+  if (context.options[context.options.length - 1] === "sonar-runtime") {
+    const encodedMessage: EncodedMessage = { secondaryLocations, message, cost };
+    reportDescriptor.message = JSON.stringify(encodedMessage);
+  }
+  context.report(reportDescriptor);
+}
+
+/**
+ * Converts `SourceLocation` range into `IssueLocation`
+ */
+export function issueLocation(
```

Comment:
I think it will be easier if this function takes nodes

Reply:
As discussed, we can keep it with SourceLocation